hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDBCheckpointDiffer.java

```diff
@@ -215,390 +288,511 @@ public static void addDebugLevel(Integer level) {
-                  }
-                }
-              }
-              if (debugEnabled(DEBUG_DAG_BUILD_UP)) {
-                printMutableGraph(null, null,
-                    compactionDAGFwd);
-              }
+    setRocksDBForCompactionTracking(rocksOptions, new ArrayList<>());
+  }
+
+  private AbstractEventListener newCompactionBeginListener() {
+    return new AbstractEventListener() {
+      @Override
+      public void onCompactionBegin(RocksDB db,
+          CompactionJobInfo compactionJobInfo) {
+
+        synchronized (compactionListenerWriteLock) {
+
+          if (compactionJobInfo.inputFiles().size() == 0) {
+            LOG.error("Compaction input files list is empty");
+            return;
+          }
+
+          // Create hardlink backups for the SST files that are going
+          // to be deleted after this RDB compaction.
+          for (String file : compactionJobInfo.inputFiles()) {
+            LOG.debug("Creating hard link for '{}'", file);
+            String saveLinkFileName =
+                sstBackupDir + new File(file).getName();
+            Path link = Paths.get(saveLinkFileName);
+            Path srcFile = Paths.get(file);
+            try {
+              Files.createLink(link, srcFile);
+            } catch (IOException e) {
+              LOG.error("Exception in creating hard link for {}", file);
+              throw new RuntimeException("Failed to create hard link", e);
             }
           }
-        };
 
-    list.add(onCompactionCompletedListener);
-    rocksOptions.setListeners(list);
+        }
+      }
+    };
   }
 
-  public RocksDB getRocksDBInstanceWithCompactionTracking(String dbPath)
-      throws RocksDBException {
-    final Options opt = new Options().setCreateIfMissing(true)
-        .setCompressionType(CompressionType.NO_COMPRESSION);
-    opt.setMaxBytesForLevelMultiplier(2);
-    setRocksDBForCompactionTracking(opt);
-    return RocksDB.open(opt, dbPath);
+  private AbstractEventListener newCompactionCompletedListener() {
+    return new AbstractEventListener() {
+      @Override
+      public void onCompactionCompleted(RocksDB db,
+          CompactionJobInfo compactionJobInfo) {
+
+        synchronized (compactionListenerWriteLock) {
+
+          if (compactionJobInfo.inputFiles().isEmpty()) {
+            LOG.error("Compaction input files list is empty");
+            return;
+          }
+
+          final StringBuilder sb = new StringBuilder();
+
+          if (LOG.isDebugEnabled()) {
+            // Print compaction reason for this entry in the log file
+            // e.g. kLevelL0FilesNum / kLevelMaxLevelSize.
+            sb.append(COMPACTION_LOG_COMMENT_LINE_PREFIX)
+                .append(compactionJobInfo.compactionReason())
+                .append('\n');
+          }
+
+          // Mark the beginning of a compaction log
+          sb.append(COMPACTION_LOG_ENTRY_LINE_PREFIX);
+
+          // Trim DB path, only keep the SST file name
```

Comment:
Can we keep the absolute path here instead of just file name. This might be useful while opening the sst files in the later steps , if file is compacted and not present in the db, we can look it up from the hardlink dir.

Reply:
The full path is highly redundant. IMO there is no need to persist the full path as the SST file can either be in the SST backup dir or in active DB dir. Storing the full path easily doubles the in-memory map memory consumption compared to file name only.